### PR TITLE
Version 0.1

### PR DESCRIPTION
**Pull request by mecha**
Initial release.

### Added
- Interop interfaces.
- Interface tests.

----

Repository to be renamed from `espresso-abstract` to `expression-abstract` after merging this PR.

### CODE REVIEW

**Comment by mecha** [comment]
This repository is to be moved from `dhii/espresso-abstract` to `dhii/expression-abstract` at the time of release.

**Review by XedinUnknown — COMMENTED** [review]
I would suggest moving the actual evaluation method out of `_evaluate()` into another, more low-level method, for even more control of evaluation behaviour.

**Comment by mecha** [comment]
@XedinUnknown Agreed.

**Comment by mecha** [comment]
@XedinUnknown Kindly see the latest commit for feedback. https://github.com/Dhii/espresso-abstract/pull/1/commits/04fdedf1eb4d68ffdc4084c6080e790469fe27fa

**Comment by mecha** [comment]
@XedinUnknown What do you think of the new namespace organization?

I'm not dead set on the `Dhii\Expression\Expression` namespace. On the one hand, everything is organized neatly; on the other hand it can be confusing since both the package and the namespace have the same name. In my opinion it's justified and not that bothering.

I'm a fan of keeping everything in its own sub namespace after the package namespace. Thoughts?

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
I think that the logic in `AbstractExpression#_setTerms()`, which validates terms, can be reduced to just `_addTerm()` throwing.

File: test/functional/Context/AbstractCompositeContextTest.php [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+            'one' => 1,
+            'two' => 2,
+        ));
+
+        $subject->this()->_setValues(array(
+            'three' => 3,
+            'four'  => 4,
+        ));
+
+        $expected = array(
```
Comment (XedinUnknown) [inline test/functional/Context/AbstractCompositeContextTest.php]: Why not call `_setValues()` with this?

File: test/functional/Expression/AbstractExpressionTest.php [hunk 2]
```diff
@@ -83,13 +82,13 @@ public function testGetTerms()
     /**
      * Tests the protected term setter method.
      *
-     * @since [*next-version*]
+     * @since 0.1
      */
     public function testSetTerms()
```
Comment (XedinUnknown) [inline test/functional/Expression/AbstractExpressionTest.php]: This tests exactly what I said not to test. Don't test properties. Only the methods.

File: test/functional/Expression/AbstractLeftAssociativeExpressionTest.php [hunk 3]
```diff
@@ -1,41 +1,42 @@
...
      *
      * @return AbstractLeftAssocOperatorExpression
      */
     public function createInstance()
     {
         $mock = $this->mock(static::TEST_SUBJECT_CLASSNAME)
-            ->_initBuffer(function() {
-                return null;
+            ->_initBuffer(function () {
+                return;
```
Comment (XedinUnknown) [inline test/functional/Expression/AbstractLeftAssociativeExpressionTest.php]: `return` is identical to `return null`. Also, it is the same as not explicitly returning anything. And that is the same as just writing `_initBuffer()`, i.e. without passing the function to it.

File: test/functional/Expression/AbstractLeftAssociativeExpressionTest.php [hunk 4]
```diff
@@ -1,41 +1,42 @@
...
+     * @since 0.1
      *
      * @return AbstractLeftAssocOperatorExpression
      */
     public function createInstance()
     {
         $mock = $this->mock(static::TEST_SUBJECT_CLASSNAME)
-            ->_initBuffer(function() {
-                return null;
+            ->_initBuffer(function () {
```
Comment (XedinUnknown) [inline test/functional/Expression/AbstractLeftAssociativeExpressionTest.php]: `return` is identical to `return null`, and is exactly what happens if you don't explicitly return anything. Also, not passing anything to mocked method will return `null` from that method, This means that you can re-write this as simply `_initBuffer()`.

File: test/functional/Expression/AbstractGenericExpressionTest.php [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+
+        $terms = array(
+            $this->mockEvaluable(),
+            $this->mockEvaluable(),
+            $this->mockEvaluable(),
+        );
+
+        $subject->setTerms($terms);
+
+        $this->assertEquals($terms, $subject->this()->terms);
```
Comment (XedinUnknown) [inline test/functional/Expression/AbstractGenericExpressionTest.php]: This does exactly what I said not to do. Please don't test properties. Test the protected getter, instead.

**Review by XedinUnknown — CHANGES_REQUESTED** [review]

**Review by XedinUnknown — CHANGES_REQUESTED** [review]
You can add the term's index to the [error message](https://github.com/Dhii/espresso-abstract/blob/release/0.1/src/Expression/AbstractExpression.php#L49) at very little cost. This can be very helpful when debugging.

**Comment by mecha** [comment]
@XedinUnknown What do you think of making the functionality outlined in [Note 2](https://github.com/Dhii/espresso-abstract/wiki/Operator-Expressions) default for right associative operators?

**Review by XedinUnknown — APPROVED** [review]
Thumbs up for removing Generic class, as well as making the requested changes. Extra credit for changing release date.

**Comment by mecha** [comment]
@XedinUnknown Made changes to the functionality of right-associative operators and updated [the docs](https://github.com/Dhii/espresso-abstract/wiki/Operator-Expressions).

**Pull request merged**